File: Qlearning_2.py
import numpy as np
import matplotlib.pyplot as plt
import heapq
from collections import deque
import random
import time

# =============================================================================
# 1. OPTIMIZED MAZE SYSTEM
# =============================================================================

class PerfectMaze:
    def __init__(self, width=15, height=15, obstacle_density=0.1):
        self.width = width
        self.height = height
        self.grid = np.zeros((height, width), dtype=np.int8)
        self.start_pos = (0, 0)
        self.end_pos = (width - 1, height - 1)
        self._generate_perfect_maze(obstacle_density)
        
        # Precompute valid actions for O(1) access
        self.valid_actions = np.empty((height, width), dtype=object)
        self._precompute_actions()

    def _generate_perfect_maze(self, density):
        """Generates a guaranteed connected maze with optimized DFS."""
        self.grid.fill(0)
        
        # Add random obstacles
        n_obstacles = int(self.width * self.height * density)
        indices = np.random.choice(self.width * self.height, n_obstacles * 2, replace=False)
        
        count = 0
        for idx in indices:
            if count >= n_obstacles:
                break
            
            y, x = divmod(idx, self.width)
            
            # Don't block start or end
            if (x, y) == self.start_pos or (x, y) == self.end_pos:
                continue
                
            self.grid[y, x] = 1
            count += 1

        # Ensure Connectivity
        self._ensure_connectivity()

    def _ensure_connectivity(self):
        """Removes obstacles if they block the path to key areas."""
        while True:
            visited = np.zeros_like(self.grid, dtype=bool)
            queue = deque([self.start_pos])
            visited[self.start_pos[1], self.start_pos[0]] = True
            reachable_count = 0
            
            while queue:
                x, y = queue.popleft()
                reachable_count += 1
                
                for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < self.width and 0 <= ny < self.height:
                        if not visited[ny, nx] and self.grid[ny, nx] == 0:
                            visited[ny, nx] = True
                            queue.append((nx, ny))
            
            # Check if all empty cells are reachable
            total_empty = np.sum(self.grid == 0)
            if reachable_count == total_empty:
                break
            
            # Identify unreachable cells
            unreachable_mask = (self.grid == 0) & (~visited)
            if not np.any(unreachable_mask):
                break
                
            # Pick a random unreachable cell
            uy, ux = np.where(unreachable_mask)
            if len(uy) > 0:
                # Remove a few random obstacles to open up paths
                obstacles = np.argwhere(self.grid == 1)
                if len(obstacles) > 0:
                    to_remove = obstacles[np.random.choice(len(obstacles), min(len(obstacles), 5), replace=False)]
                    for ry, rx in to_remove:
                        self.grid[ry, rx] = 0
            else:
                break

    def _precompute_actions(self):
        """Precomputes valid actions for every cell for O(1) lookup."""
        # Actions: 0:Right, 1:Down, 2:Left, 3:Up
        for y in range(self.height):
            for x in range(self.width):
                if self.grid[y, x] == 1:
                    self.valid_actions[y, x] = np.array([], dtype=np.int8)
                    continue
                
                actions = []
                # Right
                if x + 1 < self.width and self.grid[y, x + 1] == 0:
                    actions.append(0)
                # Down
                if y + 1 < self.height and self.grid[y + 1, x] == 0:
                    actions.append(1)
                # Left
                if x - 1 >= 0 and self.grid[y, x - 1] == 0:
                    actions.append(2)
                # Up
                if y - 1 >= 0 and self.grid[y - 1, x] == 0:
                    actions.append(3)
                
                self.valid_actions[y, x] = np.array(actions, dtype=np.int8)

    def get_possible_actions(self, x, y):
        """O(1) action lookup."""
        return self.valid_actions[y, x]

    def is_valid(self, x, y):
        return 0 <= x < self.width and 0 <= y < self.height and self.grid[y, x] == 0

# =============================================================================
# 2. OPTIMIZED SOLVER (Q-LEARNING)
# =============================================================================

class PerfectSolver:
    def __init__(self, maze, learning_rate=0.1, discount_factor=0.99, exploration_rate=1.0):
        self.maze = maze
        self.lr = learning_rate
        self.gamma = discount_factor
        self.epsilon = exploration_rate
        self.epsilon_min = 0.01
        self.epsilon_decay = 0.995
        
        # Q-Table: (y, x, action) - Optimized for fixed target (Exit)
        self.q_table = np.zeros((maze.height, maze.width, 4), dtype=np.float32)
        
        # Pre-allocate direction arrays for speed
        self.dx = np.array([1, 0, -1, 0], dtype=np.int8)
        self.dy = np.array([0, 1, 0, -1], dtype=np.int8)

    def get_action(self, x, y, tx, ty):
        """Epsilon-greedy action selection optimized."""
        possible_actions = self.maze.get_possible_actions(x, y)
        if len(possible_actions) == 0:
            return None
            
        if np.random.random() < self.epsilon:
            return np.random.choice(possible_actions)
        
        # Exploitation: Get max Q value actions
        q_values = self.q_table[y, x, possible_actions]
        max_q = np.max(q_values)
        
        best_indices = np.where(q_values == max_q)[0]
        chosen_idx = np.random.choice(best_indices)
        return possible_actions[chosen_idx]

    def update(self, x, y, action, reward, nx, ny, tx, ty):
        """Q-learning update rule."""
        current_q = self.q_table[y, x, action]
        
        # Max Q for next state
        next_actions = self.maze.get_possible_actions(nx, ny)
        if len(next_actions) > 0:
            max_next_q = np.max(self.q_table[ny, nx, next_actions])
        else:
            max_next_q = 0.0
            
        # Update
        new_q = current_q + self.lr * (reward + self.gamma * max_next_q - current_q)
        self.q_table[y, x, action] = new_q

    def decay_epsilon(self):
        self.epsilon = max(self.epsilon_min, self.epsilon * self.epsilon_decay)

# =============================================================================
# 3. HIGH-PERFORMANCE TRAINING LOOP
# =============================================================================

def train_agent(maze, agent, episodes=1000):
    print(f"Starting Training: {episodes} episodes")
    start_time = time.time()
    
    width, height = maze.width, maze.height
    dx_arr, dy_arr = agent.dx, agent.dy
    
    history_steps = []
    history_rewards = []
    
    for episode in range(episodes):
        # Select Start (Random) / Target (Fixed Exit)
        while True:
            sx, sy = np.random.randint(0, width), np.random.randint(0, height)
            tx, ty = width - 1, height - 1 # Fixed Target
            if maze.grid[sy, sx] == 0 and maze.grid[ty, tx] == 0 and (sx, sy) != (tx, ty):
                break
        
        cx, cy = sx, sy
        total_reward = 0
        steps = 0
        max_steps = width * height
        
        while steps < max_steps:
            action = agent.get_action(cx, cy, tx, ty)
            if action is None:
                break
                
            nx, ny = cx + dx_arr[action], cy + dy_arr[action]
            
            if nx == tx and ny == ty:
                reward = 100.0
                done = True
            else:
                reward = -1.0
                done = False
            
            agent.update(cx, cy, action, reward, nx, ny, tx, ty)
            
            cx, cy = nx, ny
            total_reward += reward
            steps += 1
            
            if done:
                break
        
        agent.decay_epsilon()
        history_steps.append(steps)
        history_rewards.append(total_reward)
        
        if (episode + 1) % 100 == 0:
            avg_steps = np.mean(history_steps[-100:])
            print(f"Episode {episode+1:5d} | Avg Steps: {avg_steps:6.1f} | Epsilon: {agent.epsilon:.4f}")

    duration = time.time() - start_time
    print(f"Training Complete. Time: {duration:.2f}s, Speed: {episodes/duration:.1f} eps/s")
    return history_rewards

# =============================================================================
# 4. TESTING & VISUALIZATION
# =============================================================================

def test_agent(maze, agent, start, end):
    path = [start]
    cx, cy = start
    tx, ty = end
    visited = set([start])
    
    print(f"\nTesting Path: {start} -> {end}")
    
    steps = 0
    max_steps = maze.width * maze.height * 2
    
    while (cx, cy) != (tx, ty) and steps < max_steps:
        possible_actions = maze.get_possible_actions(cx, cy)
        if not len(possible_actions):
            print("Stuck!")
            break
            
        q_values = agent.q_table[cy, cx, possible_actions]
        best_action_idx = np.argmax(q_values)
        action = possible_actions[best_action_idx]
        
        nx, ny = cx + agent.dx[action], cy + agent.dy[action]
        
        if (nx, ny) in visited:
            print("Loop detected in test path!")
            pass
            
        cx, cy = nx, ny
        path.append((cx, cy))
        visited.add((cx, cy))
        steps += 1
    
    if (cx, cy) == (tx, ty):
        print(f"Target Reached in {steps} steps.")
        return path
    else:
        print("Failed to reach target.")
        return path

def visualize_path(maze, path):
    plt.figure(figsize=(8, 8))
    plt.imshow(maze.grid, cmap='binary')
    
    if path:
        px, py = zip(*path)
        plt.plot(px, py, color='red', linewidth=3, marker='o', markersize=4, alpha=0.7)
        plt.plot(px[0], py[0], 'go', markersize=10, label='Start')
        plt.plot(px[-1], py[-1], 'bx', markersize=10, label='End')
    
    plt.title("Optimized Q-Learning Path")
    plt.legend()
    plt.grid(False)
    plt.savefig('result.png')
    print("Plot saved to result.png")

if __name__ == "__main__":
    # Configuration
    SIZE = 15
    DENSITY = 0.15
    EPISODES = 5000
    
    # Initialize
    maze = PerfectMaze(SIZE, SIZE, DENSITY)
    agent = PerfectSolver(maze, learning_rate=0.2, discount_factor=0.95)
    
    # Train
    train_agent(maze, agent, EPISODES)
    
    # Test
    start = (0, 0)
    end = (SIZE-1, SIZE-1)
    
    # Ensure start/end are valid
    if maze.grid[start[1], start[0]] == 1: maze.grid[start[1], start[0]] = 0
    if maze.grid[end[1], end[0]] == 1: maze.grid[end[1], end[0]] = 0
    
    path = test_agent(maze, agent, start, end)
    visualize_path(maze, path)

File: Qlearning_2.cpp

#include <iostream>
#include <vector>
#include <random>
#include <cmath>
#include <algorithm>
#include <deque>
#include <fstream>
#include <string>
#include <chrono>
#include <iomanip>
#include <set>

// =============================================================================
// 0. UTILS
// =============================================================================

// Simple struct to represent a point
struct Point {
    int x;
    int y;
    
    bool operator==(const Point& other) const {
        return x == other.x && y == other.y;
    }
    
    bool operator!=(const Point& other) const {
        return !(*this == other);
    }
    
    // For using Point in std::set/map
    bool operator<(const Point& other) const {
        if (x != other.x) return x < other.x;
        return y < other.y;
    }
};

// PPM Image Writer to replace Matplotlib
void save_ppm(const std::string& filename, int width, int height, 
              const std::vector<std::vector<int>>& grid, 
              const std::vector<Point>& path) {
    std::ofstream ofs(filename);
    if (!ofs) {
        std::cerr << "Error: Could not open " << filename << " for writing." << std::endl;
        return;
    }

    // P3 format (ASCII RGB)
    ofs << "P3\n" << width * 20 << " " << height * 20 << "\n255\n";

    // Create a visual grid (scaling up by 20x for visibility)
    // 0 = Empty (White), 1 = Obstacle (Black)
    // Path = Red
    // Start = Green, End = Blue

    std::set<Point> path_set(path.begin(), path.end());
    Point start = path.front();
    Point end = path.back();

    for (int y = 0; y < height * 20; ++y) {
        for (int x = 0; x < width * 20; ++x) {
            int grid_x = x / 20;
            int grid_y = y / 20;
            
            Point p = {grid_x, grid_y};
            
            int r, g, b;
            
            if (p == start) {
                r = 0; g = 255; b = 0; // Green
            } else if (p == end) {
                r = 0; g = 0; b = 255; // Blue
            } else if (path_set.count(p)) {
                 r = 255; g = 0; b = 0; // Red
            } else if (grid[grid_y][grid_x] == 1) {
                r = 0; g = 0; b = 0; // Black (Obstacle)
            } else {
                r = 255; g = 255; b = 255; // White (Empty)
            }
            
            ofs << r << " " << g << " " << b << " ";
        }
        ofs << "\n";
    }
    
    std::cout << "Plot saved to " << filename << " (PPM format)" << std::endl;
}

// =============================================================================
// 1. OPTIMIZED MAZE SYSTEM
// =============================================================================

class PerfectMaze {
public:
    int width;
    int height;
    std::vector<std::vector<int>> grid;
    Point start_pos;
    Point end_pos;
    // valid_actions[y][x] contains list of actions (0:Right, 1:Down, 2:Left, 3:Up)
    std::vector<std::vector<std::vector<int>>> valid_actions;

    PerfectMaze(int w = 15, int h = 15, double obstacle_density = 0.1) 
        : width(w), height(h) {
        grid.resize(height, std::vector<int>(width, 0));
        start_pos = {0, 0};
        end_pos = {width - 1, height - 1};
        valid_actions.resize(height, std::vector<std::vector<int>>(width));
        
        generate_perfect_maze(obstacle_density);
        precompute_actions();
    }

    const std::vector<int>& get_possible_actions(int x, int y) const {
        return valid_actions[y][x];
    }

    bool is_valid(int x, int y) const {
        return x >= 0 && x < width && y >= 0 && y < height && grid[y][x] == 0;
    }

private:
    void generate_perfect_maze(double density) {
        // Clear grid
        for(auto& row : grid) std::fill(row.begin(), row.end(), 0);

        int n_obstacles = static_cast<int>(width * height * density);
        std::vector<int> indices(width * height);
        for(int i=0; i < width * height; ++i) indices[i] = i;

        // Shuffle indices
        unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
        std::mt19937 g(seed);
        std::shuffle(indices.begin(), indices.end(), g);

        int count = 0;
        for (int idx : indices) {
            if (count >= n_obstacles) break;
            
            int y = idx / width;
            int x = idx % width;
            
            if ((x == start_pos.x && y == start_pos.y) || (x == end_pos.x && y == end_pos.y)) {
                continue;
            }
            
            grid[y][x] = 1;
            count++;
        }

        ensure_connectivity();
    }

    void ensure_connectivity() {
        while (true) {
            std::vector<std::vector<bool>> visited(height, std::vector<bool>(width, false));
            std::deque<Point> queue;
            
            queue.push_back(start_pos);
            visited[start_pos.y][start_pos.x] = true;
            
            int reachable_count = 0;
            
            int dx[] = {0, 1, 0, -1};
            int dy[] = {1, 0, -1, 0};
            
            while (!queue.empty()) {
                Point p = queue.front();
                queue.pop_front();
                reachable_count++;
                
                for (int i = 0; i < 4; ++i) {
                    int nx = p.x + dx[i];
                    int ny = p.y + dy[i];
                    
                    if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
                        if (!visited[ny][nx] && grid[ny][nx] == 0) {
                            visited[ny][nx] = true;
                            queue.push_back({nx, ny});
                        }
                    }
                }
            }
            
            int total_empty = 0;
            for(const auto& row : grid) {
                for(int cell : row) {
                    if (cell == 0) total_empty++;
                }
            }
            
            if (reachable_count == total_empty) break;
            
            // Find unreachable cells
            std::vector<Point> unreachable;
            for (int y = 0; y < height; ++y) {
                for (int x = 0; x < width; ++x) {
                    if (grid[y][x] == 0 && !visited[y][x]) {
                        unreachable.push_back({x, y});
                    }
                }
            }
            
            if (unreachable.empty()) break;
            
            // Remove random obstacles
            std::vector<Point> obstacles;
            for (int y = 0; y < height; ++y) {
                for (int x = 0; x < width; ++x) {
                    if (grid[y][x] == 1) obstacles.push_back({x, y});
                }
            }
            
            if (!obstacles.empty()) {
                unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
                std::mt19937 g(seed);
                std::shuffle(obstacles.begin(), obstacles.end(), g);
                
                int remove_count = std::min((int)obstacles.size(), 5);
                for (int i = 0; i < remove_count; ++i) {
                    grid[obstacles[i].y][obstacles[i].x] = 0;
                }
            } else {
                break;
            }
        }
    }

    void precompute_actions() {
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                if (grid[y][x] == 1) continue;
                
                std::vector<int> actions;
                // 0:Right, 1:Down, 2:Left, 3:Up
                // Right
                if (x + 1 < width && grid[y][x + 1] == 0) actions.push_back(0);
                // Down
                if (y + 1 < height && grid[y + 1][x] == 0) actions.push_back(1);
                // Left
                if (x - 1 >= 0 && grid[y][x - 1] == 0) actions.push_back(2);
                // Up
                if (y - 1 >= 0 && grid[y - 1][x] == 0) actions.push_back(3);
                
                valid_actions[y][x] = actions;
            }
        }
    }
};

// =============================================================================
// 2. OPTIMIZED SOLVER (Q-LEARNING)
// =============================================================================

class PerfectSolver {
public:
    PerfectMaze& maze;
    float lr;
    float gamma;
    float epsilon;
    float epsilon_min;
    float epsilon_decay;
    
    // q_table[y][x][action]
    std::vector<std::vector<std::vector<float>>> q_table;
    
    // dx/dy for actions 0, 1, 2, 3
    const int dx[4] = {1, 0, -1, 0};
    const int dy[4] = {0, 1, 0, -1};

    std::mt19937 rng;

    PerfectSolver(PerfectMaze& m, float learning_rate = 0.1, float discount_factor = 0.99, float exploration_rate = 1.0)
        : maze(m), lr(learning_rate), gamma(discount_factor), epsilon(exploration_rate) {
        
        epsilon_min = 0.01f;
        epsilon_decay = 0.995f;
        
        q_table.resize(maze.height, std::vector<std::vector<float>>(maze.width, std::vector<float>(4, 0.0f)));
        
        unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
        rng.seed(seed);
    }

    int get_action(int x, int y, int tx, int ty) {
        const auto& possible_actions = maze.get_possible_actions(x, y);
        if (possible_actions.empty()) return -1;
        
        std::uniform_real_distribution<float> dist(0.0, 1.0);
        
        if (dist(rng) < epsilon) {
            std::uniform_int_distribution<int> action_dist(0, possible_actions.size() - 1);
            return possible_actions[action_dist(rng)];
        }
        
        float max_q = -1e9;
        std::vector<int> best_actions;
        
        for (int action : possible_actions) {
            float q = q_table[y][x][action];
            if (q > max_q) {
                max_q = q;
                best_actions.clear();
                best_actions.push_back(action);
            } else if (std::abs(q - max_q) < 1e-6) {
                best_actions.push_back(action);
            }
        }
        
        if (best_actions.empty()) return possible_actions[0]; // Fallback
        
        std::uniform_int_distribution<int> best_dist(0, best_actions.size() - 1);
        return best_actions[best_dist(rng)];
    }

    void update(int x, int y, int action, float reward, int nx, int ny, int tx, int ty) {
        float current_q = q_table[y][x][action];
        float max_next_q = 0.0f;
        
        const auto& next_actions = maze.get_possible_actions(nx, ny);
        if (!next_actions.empty()) {
            max_next_q = -1e9;
            for (int na : next_actions) {
                float q = q_table[ny][nx][na];
                if (q > max_next_q) max_next_q = q;
            }
        } else {
             // If no actions, q is 0 (or terminal)
             max_next_q = 0.0f;
        }
        
        // if max_next_q is still initial small value, set to 0
        if (max_next_q == -1e9) max_next_q = 0.0f;

        float new_q = current_q + lr * (reward + gamma * max_next_q - current_q);
        q_table[y][x][action] = new_q;
    }

    void decay_epsilon() {
        epsilon = std::max(epsilon_min, epsilon * epsilon_decay);
    }
};

// =============================================================================
// 3. HIGH-PERFORMANCE TRAINING LOOP
// =============================================================================

std::vector<float> train_agent(PerfectMaze& maze, PerfectSolver& agent, int episodes = 1000) {
    std::cout << "Starting Training: " << episodes << " episodes" << std::endl;
    auto start_time = std::chrono::high_resolution_clock::now();
    
    std::vector<float> history_rewards;
    std::vector<int> history_steps;
    
    int width = maze.width;
    int height = maze.height;
    
    std::mt19937 rng;
    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    rng.seed(seed);
    std::uniform_int_distribution<int> w_dist(0, width - 1);
    std::uniform_int_distribution<int> h_dist(0, height - 1);

    for (int episode = 0; episode < episodes; ++episode) {
        int sx, sy, tx, ty;
        while (true) {
            sx = w_dist(rng);
            sy = h_dist(rng);
            tx = width - 1;
            ty = height - 1;
            
            if (maze.grid[sy][sx] == 0 && maze.grid[ty][tx] == 0 && (sx != tx || sy != ty)) {
                break;
            }
        }
        
        int cx = sx;
        int cy = sy;
        float total_reward = 0;
        int steps = 0;
        int max_steps = width * height;
        bool done = false;
        
        while (steps < max_steps) {
            int action = agent.get_action(cx, cy, tx, ty);
            if (action == -1) break;
            
            int nx = cx + agent.dx[action];
            int ny = cy + agent.dy[action];
            
            float reward;
            if (nx == tx && ny == ty) {
                reward = 100.0f;
                done = true;
            } else {
                reward = -1.0f;
                done = false;
            }
            
            agent.update(cx, cy, action, reward, nx, ny, tx, ty);
            
            cx = nx;
            cy = ny;
            total_reward += reward;
            steps++;
            
            if (done) break;
        }
        
        agent.decay_epsilon();
        history_steps.push_back(steps);
        history_rewards.push_back(total_reward);
        
        if ((episode + 1) % 100 == 0) {
            float avg_steps = 0;
            int count = 0;
            for (int i = std::max(0, (int)history_steps.size() - 100); i < history_steps.size(); ++i) {
                avg_steps += history_steps[i];
                count++;
            }
            if (count > 0) avg_steps /= count;
            
            std::cout << "Episode " << std::setw(5) << episode + 1 
                      << " | Avg Steps: " << std::fixed << std::setprecision(1) << avg_steps 
                      << " | Epsilon: " << std::setprecision(4) << agent.epsilon << std::endl;
        }
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duration = end_time - start_time;
    
    std::cout << "Training Complete. Time: " << duration.count() << "s, Speed: " 
              << episodes / duration.count() << " eps/s" << std::endl;
              
    return history_rewards;
}

// =============================================================================
// 4. TESTING & VISUALIZATION
// =============================================================================

std::vector<Point> test_agent(PerfectMaze& maze, PerfectSolver& agent, Point start, Point end) {
    std::vector<Point> path;
    path.push_back(start);
    
    int cx = start.x;
    int cy = start.y;
    int tx = end.x;
    int ty = end.y;
    
    std::set<Point> visited;
    visited.insert(start);
    
    std::cout << "\nTesting Path: (" << start.x << "," << start.y << ") -> (" << end.x << "," << end.y << ")" << std::endl;
    
    int steps = 0;
    int max_steps = maze.width * maze.height * 2;
    
    while ((cx != tx || cy != ty) && steps < max_steps) {
        const auto& possible_actions = maze.get_possible_actions(cx, cy);
        if (possible_actions.empty()) {
            std::cout << "Stuck!" << std::endl;
            break;
        }
        
        int best_action = -1;
        float max_q = -1e9;
        
        for (int action : possible_actions) {
            float q = agent.q_table[cy][cx][action];
            if (q > max_q) {
                max_q = q;
                best_action = action;
            }
        }
        
        if (best_action == -1) break;
        
        int nx = cx + agent.dx[best_action];
        int ny = cy + agent.dy[best_action];
        
        Point np = {nx, ny};
        if (visited.count(np)) {
            std::cout << "Loop detected in test path!" << std::endl;
            // Depending on strictness, we might break or continue.
            // Python code says 'pass', so loop continues.
        }
        
        cx = nx;
        cy = ny;
        path.push_back({cx, cy});
        visited.insert({cx, cy});
        steps++;
    }
    
    if (cx == tx && cy == ty) {
        std::cout << "Target Reached in " << steps << " steps." << std::endl;
    } else {
        std::cout << "Failed to reach target." << std::endl;
    }
    
    return path;
}

int main() {
    // Configuration
    const int SIZE = 15;
    const double DENSITY = 0.15;
    const int EPISODES = 5000;
    
    // Initialize
    PerfectMaze maze(SIZE, SIZE, DENSITY);
    PerfectSolver agent(maze, 0.2f, 0.95f);
    
    // Train
    train_agent(maze, agent, EPISODES);
    
    // Test
    Point start = {0, 0};
    Point end = {SIZE - 1, SIZE - 1};
    
    // Ensure start/end are valid (clearing obstacles)
    if (maze.grid[start.y][start.x] == 1) maze.grid[start.y][start.x] = 0;
    if (maze.grid[end.y][end.x] == 1) maze.grid[end.y][end.x] = 0;
    
    // Re-verify connectivity/actions after forcing start/end 
    // (In Python code this was just a direct set, simple fix)
    
    std::vector<Point> path = test_agent(maze, agent, start, end);
    save_ppm("result.ppm", SIZE, SIZE, maze.grid, path);
    
    return 0;
}
